src/libmeasurement_kit/ooni/utils.cpp

```diff
@@ -23,141 +23,103 @@ void resolver_lookup(Callback<Error, std::string> callback, Settings settings,
     resolver_lookup_impl(callback, settings, reactor, logger);
 }
 
-IPLocation::IPLocation(std::string path_country_, std::string path_asn_,
-                       std::string path_city_) {
-    path_asn = path_asn_;
-    path_country = path_country_;
-    path_city = path_city_;
+/* static */ Var<GeoipCache> GeoipCache::global() {
+    static Var<GeoipCache> singleton(new GeoipCache);
+    return singleton;
 }
 
-IPLocation::~IPLocation() {
-    if (gi_asn != nullptr) {
-        GeoIP_delete(gi_asn);
+Var<GeoipDatabase> GeoipCache::get(std::string path, bool &did_open) {
+    if (instances.find(path) != instances.end()) {
+        did_open = false;
+        return instances.at(path);
     }
-    if (gi_city != nullptr) {
-        GeoIP_delete(gi_city);
-    }
-    if (gi_country != nullptr) {
-        GeoIP_delete(gi_country);
+    did_open = true;
+    if (instances.size() > max_size) {
+        instances.erase(std::prev(instances.end()));
     }
+    instances[path] = Var<GeoipDatabase>(new GeoipDatabase(path));
+    return instances[path];
 }
 
-ErrorOr<std::string> IPLocation::resolve_country_code(std::string ip,
-                                                      Var<Logger> logger) {
-    if (gi_country == nullptr) {
-        gi_country = GeoIP_open(path_country.c_str(), GEOIP_MEMORY_CACHE);
-        if (gi_country == nullptr) {
-            logger->warn("IPLocation: cannot open geoip country database");
-            return CannotOpenGeoIpCountryDatabaseError();
+ErrorOr<std::string> GeoipDatabase::with_open_database_do(
+        std::function<ErrorOr<std::string>()> action,
+        Var<Logger> logger) {
+    if (!db) {
+        db.reset(GeoIP_open(path.c_str(), GEOIP_MEMORY_CACHE),
+                 [](GeoIP *pointer) {
+                    if (pointer) {
+                        GeoIP_delete(pointer);
+                    }
+                 });
+        if (!db) {
+            logger->warn("cannot open geoip database: %s", path.c_str());
+            return GeoipDatabaseOpenError();
         }
+        // FALLTHROUGH
     }
-    GeoIPLookup gl;
-    memset(&gl, 0, sizeof(gl));
-
-    const char *result;
-    result = GeoIP_country_code_by_name_gl(gi_country, ip.c_str(), &gl);
-    if (result == nullptr) {
-        return GenericError();
-    }
-    std::string country_code = result;
-    return country_code;
+    return action();
 }
 
-ErrorOr<std::string> IPLocation::resolve_country_name(std::string ip,
-                                                      Var<Logger> logger) {
-    if (gi_country == nullptr) {
-        gi_country = GeoIP_open(path_country.c_str(), GEOIP_MEMORY_CACHE);
-        if (gi_country == nullptr) {
-            logger->warn("IPLocation: cannot open geoip country database");
-            return CannotOpenGeoIpCountryDatabaseError();
```

Comment:
This code that opens the database has been refactored inside the method called `with_open_database_do`